rados/snapshot.go

```diff
@@ -0,0 +1,131 @@
+	return GetRadosError(int(ret))
+}
+
+// SnapRemove deletes a pool snapshot
+func (ioctx *IOContext) SnapRemove(snapname string) error {
+	cSnapname := C.CString(snapname)
+	defer C.free(unsafe.Pointer(cSnapname))
+
+	ret := C.rados_ioctx_snap_remove(ioctx.ioctx, cSnapname)
+	return GetRadosError(int(ret))
+}
+
+// SnapRollback an object to a pool snapshot
+// The contents of the object will be the same as when the snapshot was taken.
+func (ioctx *IOContext) SnapRollback(oid, snapname string) error {
+	cOid := C.CString(oid)
+	defer C.free(unsafe.Pointer(cOid))
+
+	cSnapname := C.CString(snapname)
+	defer C.free(unsafe.Pointer(cSnapname))
+
+	ret := C.rados_ioctx_snap_rollback(ioctx.ioctx, cOid, cSnapname)
+	return GetRadosError(int(ret))
+}
+
+// SetSnapRead sets the snapshot from which reads are performed.
+// Subsequent reads will return data as it was at the time of that snapshot.
+// Specify the id of the snapshot to set or use SnapHead for no snapshot in order to
+// resume normal operation on the ioctx.
+func (ioctx *IOContext) SetSnapRead(snap uint64) {
+	C.rados_ioctx_snap_set_read(ioctx.ioctx, C.rados_snap_t(snap))
+}
+
+// SnapList returns the ids of pool snapshots.
+// Specify the maximum length of the returned array (512 by default if zero or negative)
+// If the number of snapshots is greater than this length then -ERANGE is returned and
+// the user should retry with a larger maxlen.
+func (ioctx *IOContext) SnapList(maxlen int) ([]uint64, error) {
+	if maxlen < 1 {
+		maxlen = 512
+	}
+
+	snaps := make([]uint64, maxlen)
+	ret := int(C.rados_ioctx_snap_list(ioctx.ioctx, (*C.rados_snap_t)(unsafe.Pointer(&snaps[0])), C.int(maxlen)))
+	if ret < 0 {
+		return nil, GetRadosError(ret)
+	}
+	return snaps[0:ret], nil
+}
+
+// SnapLookup gets the id of a pool snapshot
+func (ioctx *IOContext) SnapLookup(snapname string) (uint64, error) {
+	cSnapname := C.CString(snapname)
+	defer C.free(unsafe.Pointer(cSnapname))
+
+	var snap uint64
+	ret := int(C.rados_ioctx_snap_lookup(
+		ioctx.ioctx, cSnapname, (*C.rados_snap_t)(unsafe.Pointer(&snap)),
+	))
+
+	if ret == 0 {
+		return snap, nil
+	}
+	return 0, GetRadosError(ret)
+}
+
+// SnapGetName gets the name of a pool snapshot
+func (ioctx *IOContext) SnapGetName(snap uint64) (string, error) {
+	buf := make([]byte, 4096)
+	ret := int(C.rados_ioctx_snap_get_name(
+		ioctx.ioctx, C.rados_snap_t(snap), (*C.char)(unsafe.Pointer(&buf[0])), C.int(len(buf)),
+	))
+
+	if ret == 0 {
+		value := C.GoString((*C.char)(unsafe.Pointer(&buf[0])))
+		return value, nil
+	}
+
+	return "", GetRadosError(ret)
+}
+
+// SnapGetStamp finds when a pool snapshort occurred
+func (ioctx *IOContext) SnapGetStamp(snap uint64) (time.Time, error) {
+	var nsec int64
+	ret := int(C.rados_ioctx_snap_get_stamp(
+		ioctx.ioctx, C.rados_snap_t(snap), (*C.time_t)(unsafe.Pointer(&nsec)),
+	))
+
+	if ret == 0 {
+		// Handle success case
+		return time.Unix(0, nsec), nil
+	}
+	return time.Time{}, GetRadosError(ret)
```

Comment:
ah, the old "what to return on error with time struct" problem

Reply:
so true